### ISSUE TITLE
"Invalid redirect parameter" error with v0.0.4 and latest

### ISSUE THREAD

**Issue by victornoel** [pomerium/pomerium#152]
**Describe the bug**

When I try to connect to pomerium protected service, I get redirected to the authenticate service public url and the following error is shown:
![image](https://user-images.githubusercontent.com/160975/58718460-dbc27780-83cd-11e9-885f-4e5c6e8d5bd5.png)

**To Reproduce**

I used the helm chart from pomerium/pomerium-helm@1b999f445c8f5e7f361345e310ee5be28abf9119 but applied the few changes needed for v0.0.4 to work (see https://github.com/pomerium/pomerium/releases/tag/v0.0.4).

Edit: I also tried with latest chart and helm, cf comments below.

I'm using the Google IdP and a service account.

I think the configuration is the same as when it was working before with v0.0.3.

**Expected behavior**

I would expect at least to understand what is going wrong, I guess maybe I made an error somewhere but there is zero feedback.

**Environment:**

- Pomerium version (retrieve with `pomerium --version`): v0.0.4+4bdb917, and also latest
- Server Operating System/Architecture/Cloud: AKS/K8S/Helm

Configuration file(s):

```text
    - from: app.xxx.com
      to: http://my-app.my-app.svc.cluster.local:80
      allowed_domains:
      - xxx.com
      timeout: 10m
```

Logs(s):

```text
:56PM DBG authenticate: request duration=0.435802 fwd_ip=10.240.0.35 ip=10.240.0.61 method=GET req_id=a84d3895-836b-9d2a-95ca-af51844c6722 size=1950 status=400 url=/sign_in?redirect_uri=https%3A%2F%2Fapp.xxx.com%2F.pomerium%2Fcallback&response_type=code&shared_secret=ZGhZZ0XXXXXKMEMrbTAwSDZ5TVBNNFRkZko5dnc%3D&sig=2ePoPCSkUgh5pFnGU-zXvevHjKHYB7w5jHf8Kt5dbUk%3D&state=pl3hQJZwffs5lmcpqdVkYljWgDkyrK0F2d5X6oUErWcFofvytP8yhESXB8YqhGJB-CgBLNbeoHBRGKexxLqSWp21BPgGo1y45uXf3FPHYfKTNKfT5XAN88-eQT88b557ho8yCyIcd7qX2HrWohuWZHJSDn8KiPJ3Hh_z3vruDwwIQsYnVHc0m88cYGKPtVjxFTkfjrTH8w&ts=1559318162 user_agent="Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0"
```

**Comment by desimone** [comment]
Sorry about that @victornoel ; do you mind posting your configuration?

**Comment by desimone** [comment]
Almost positive I know what this is, I hit the same thing (#148).

Redirect URL is being built from authenticateURL

```go
	redirectURL := opts.AuthenticateURL
	redirectURL.Path = "/oauth2/callback"

```
RedirectURI validation uses the redirect url

```go
		middleware.ValidateRedirectURI(a.RedirectURL))
```

and makes sure they both originate from the same subdomain

```go
// ValidateRedirectURI checks the redirect uri in the query parameters and ensures that
// the its domain is in the list of proxy root domains.
func ValidateRedirectURI(rootDomain *url.URL) func(next http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			err := r.ParseForm()
			if err != nil {
				httputil.ErrorResponse(w, r, err.Error(), http.StatusBadRequest)
				return
			}
			redirectURI, err := url.Parse(r.Form.Get("redirect_uri"))
			if err != nil {
				httputil.ErrorResponse(w, r, err.Error(), http.StatusBadRequest)
				return
			}
			if !SameSubdomain(redirectURI, rootDomain) {
				httputil.ErrorResponse(w, r, "Invalid redirect parameter", http.StatusBadRequest)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}
```

So the fix should would be to set AuthenticateURL (`AUTHENTICATE_SERVICE_URL`) visible to the authenticate service. What made this bug even more confusing and frustrating, as you point out, was there was no feedback. Authenticate service should never be able to run without that setting.

It's on the docs but should be added to the `UPGRADING.md` guide.

**Comment by victornoel** [comment]
@desimone good find, I will test this on monday, thanks

**Comment by victornoel** [comment]
@desimone actually, it doesn't seem to solve my problem, so I tried with latest pomerium and https://github.com/pomerium/pomerium-helm/pull/5 and I get the same error.

I will share my config soon.

**Comment by victornoel** [comment]
@desimone here is the kubernetes resources of my deployment as outputted by helm (I have my own chart that depends on pomerium-helm where I setup the ingresses, config and secrets.)

Note that all the secrets related to TLS are those generated by the pomerium-helm chart (via `generateTLS`) and that I persisted to reuse in my sealed secret (which is properly transformed to a secret that the pomerium pods can read without problem).

```yaml
# Source: pomerium-sso/templates/configs.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: pomerium
  labels:    
    app.kubernetes.io/name: pomerium-sso
    app.kubernetes.io/instance: pomerium
    helm.sh/chart: pomerium-sso-1.0.0
    app.kubernetes.io/managed-by: Tiller
data:
  config.yaml: |-
    policy:
      - from: app.xxx.com
        to: http://my-app.my-app.svc.cluster.local:80
        allowed_domains:
        - xxx.com
        timeout: 10m
        

---
# Source: pomerium-sso/charts/pomerium/templates/authenticate-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: pomerium-authenticate
  labels:
    app.kubernetes.io/name: pomerium-authenticate
    helm.sh/chart: pomerium-1.0.0
    app.kubernetes.io/instance: pomerium
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/component: authenticate
spec:
  type: ClusterIP
  ports:
    - port: 443
      targetPort: https
      protocol: TCP
      name: https
  selector:
    app.kubernetes.io/name: pomerium-authenticate
    app.kubernetes.io/instance: pomerium

---
# Source: pomerium-sso/charts/pomerium/templates/authorize-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: pomerium-authorize
  labels:
    app.kubernetes.io/name: pomerium-authorize
    helm.sh/chart: pomerium-1.0.0
    app.kubernetes.io/instance: pomerium
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/component: authorize
spec:
  type: ClusterIP
  ports:
    - port: 443
      targetPort: https
      protocol: TCP
      name: https
  selector:
    app.kubernetes.io/name: pomerium-authorize
    app.kubernetes.io/instance: pomerium

---
# Source: pomerium-sso/charts/pomerium/templates/proxy-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: pomerium-proxy
  labels:
    app.kubernetes.io/name: pomerium-proxy
    helm.sh/chart: pomerium-1.0.0
    app.kubernetes.io/instance: pomerium
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/component: proxy
spec:
  type: ClusterIP
  ports:
    - port: 443
      targetPort: https
      protocol: TCP
      name: https
  selector:
    app.kubernetes.io/name: pomerium-proxy
    app.kubernetes.io/instance: pomerium

---
# Source: pomerium-sso/charts/pomerium/templates/authenticate-deployment.yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: pomerium-authenticate
    helm.sh/chart: pomerium-1.0.0
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/instance: pomerium
    app.kubernetes.io/component: authenticate
  name: pomerium-authenticate
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: pomerium-authenticate
      app.kubernetes.io/instance: pomerium
  template:
    metadata:
      annotations:
        config/checksum: 52133ddfcd583280388a8bbfb67a417b6b660f75bc63ec1b01beec811d4498ef
      labels:
        app.kubernetes.io/name: pomerium-authenticate
        helm.sh/chart: pomerium-1.0.0
        app.kubernetes.io/managed-by: Tiller
        app.kubernetes.io/instance: pomerium
    spec:
      containers:
      - name: pomerium
        image: pomerium/pomerium:latest
        imagePullPolicy: Always
        args:
          - --config=/etc/pomerium/config.yaml
        env:
        - name: SERVICES
          value: authenticate
        - name: AUTHENTICATE_SERVICE_URL
          value: https://auth.aks.xxx.com
        - name: COOKIE_SECRET
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: cookie-secret
        - name: SHARED_SECRET
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: shared-secret
        - name: IDP_PROVIDER
          value: google
        - name: IDP_CLIENT_ID
          valueFrom:
            secretKeyRef:
              name:  pomerium
              key: idp-client-id
        - name: IDP_CLIENT_SECRET
          valueFrom:
            secretKeyRef:
              name:  pomerium
              key: idp-client-secret
        - name: IDP_PROVIDER_URL
          value: 
        - name: CERTIFICATE
          valueFrom:
            secretKeyRef:
              name:  pomerium
              key:  authenticate-cert
        - name: CERTIFICATE_KEY
          valueFrom:
            secretKeyRef:
              name:  pomerium
              key:  authenticate-key
        - name: CERTIFICATE_AUTHORITY
          valueFrom:
            secretKeyRef:
              name:  pomerium
              key:  ca-cert
        - name: DEFAULT_UPSTREAM_TIMEOUT
          value: "5m"
        ports:
          - containerPort: 443
            name: https
            protocol: TCP
        livenessProbe:
          httpGet:
            path: /ping
            port: https
            scheme: HTTPS
        readinessProbe:
          httpGet:
            path: /ping
            port: https
            scheme: HTTPS
        volumeMounts:
        - mountPath: /etc/pomerium/
          name: config
      volumes:
      - name: config
        configMap:
          name: pomerium
---
# Source: pomerium-sso/charts/pomerium/templates/authorize-deployment.yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: pomerium-authorize
    helm.sh/chart: pomerium-1.0.0
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/instance: pomerium
    app.kubernetes.io/component: authorize
  name: pomerium-authorize
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: pomerium-authorize
      app.kubernetes.io/instance: pomerium
  template:
    metadata:
      annotations:
        config/checksum: 52133ddfcd583280388a8bbfb67a417b6b660f75bc63ec1b01beec811d4498ef
      labels:
        app.kubernetes.io/name: pomerium-authorize
        helm.sh/chart: pomerium-1.0.0
        app.kubernetes.io/managed-by: Tiller
        app.kubernetes.io/instance: pomerium
    spec:
      containers:
      - name: pomerium
        image: pomerium/pomerium:latest
        imagePullPolicy: Always
        args:
          - --config=/etc/pomerium/config.yaml
        env:
        - name: SERVICES
          value: authorize
        - name: SHARED_SECRET
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: shared-secret
        - name: CERTIFICATE
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: authorize-cert
        - name: CERTIFICATE_KEY
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: authorize-key
        - name: CERTIFICATE_AUTHORITY
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: ca-cert
        - name: DEFAULT_UPSTREAM_TIMEOUT
          value: "5m"
        ports:
          - containerPort: 443
            name: https
            protocol: TCP
        livenessProbe:
          httpGet:
            path: /ping
            port: https
            scheme: HTTPS
        readinessProbe:
          httpGet:
            path: /ping
            port: https
            scheme: HTTPS
        volumeMounts:
        - mountPath: /etc/pomerium/
          name: config
      volumes:
      - name: config
        configMap:
          name: pomerium

---
# Source: pomerium-sso/charts/pomerium/templates/proxy-deployment.yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: pomerium-proxy
    helm.sh/chart: pomerium-1.0.0
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/instance: pomerium
    app.kubernetes.io/component: proxy
  name: pomerium-proxy
spec:
  replicas: 1
  selector:
    matchLabels:  
      app.kubernetes.io/name: pomerium-proxy
      app.kubernetes.io/instance: pomerium
  template:
    metadata:
      annotations:
        config/checksum: 52133ddfcd583280388a8bbfb67a417b6b660f75bc63ec1b01beec811d4498ef
      labels:
        app.kubernetes.io/name: pomerium-proxy
        helm.sh/chart: pomerium-1.0.0
        app.kubernetes.io/managed-by: Tiller
        app.kubernetes.io/instance: pomerium
    spec:
      containers:
      - name: pomerium
        image: pomerium/pomerium:latest
        imagePullPolicy: Always
        args:
          - --config=/etc/pomerium/config.yaml
        env:
        - name: SERVICES
          value: proxy
        - name: COOKIE_SECRET
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: cookie-secret
        - name: SHARED_SECRET
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: shared-secret
        - name: AUTHENTICATE_SERVICE_URL
          value: https://auth.aks.xxx.com
        - name: AUTHORIZE_SERVICE_URL
          value: https://pomerium-authorize.proxy.svc.cluster.local
        - name: AUTHENTICATE_INTERNAL_URL
          value: https://pomerium-authenticate.proxy.svc.cluster.local
        - name: CERTIFICATE
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: proxy-cert
        - name: CERTIFICATE_KEY
          valueFrom:
            secretKeyRef:
              name: pomerium
              key: proxy-key
        - name: CERTIFICATE_AUTHORITY
          valueFrom:
            secretKeyRef:
              name: pomerium
              key:  ca-cert
        - name: DEFAULT_UPSTREAM_TIMEOUT
          value: "5m"
        ports:
          - containerPort: 443
            name: https
            protocol: TCP
        livenessProbe:
          httpGet:
            path: /ping
            port: https
            scheme: HTTPS
        readinessProbe:
          httpGet:
            path: /ping
            port: https
            scheme: HTTPS
        volumeMounts:
        - mountPath: /etc/pomerium/
          name: config
      volumes:
      - name: config
        configMap:
          name: pomerium
---
# Source: pomerium-sso/templates/ingresses.yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: pomerium-pomerium-sso-authenticate
  labels:    
    app.kubernetes.io/name: pomerium-sso
    app.kubernetes.io/instance: pomerium
    helm.sh/chart: pomerium-sso-1.0.0
    app.kubernetes.io/managed-by: Tiller
  annotations:
    certmanager.k8s.io/cluster-issuer: "letsencrypt-prod"
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
    nginx.ingress.kubernetes.io/proxy-buffer-size: "16k"
spec:
  rules:
    - host: "auth.aks.xxx.com"
      http:
        paths:
          - backend:
              serviceName: pomerium-authenticate
              servicePort: https
  tls:
    - secretName: tls-prod-auth.aks
      hosts:
        - "auth.aks.xxx.com"
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: pomerium-pomerium-sso-app.xxx.com
  labels:    
    app.kubernetes.io/name: pomerium-sso
    app.kubernetes.io/instance: pomerium
    helm.sh/chart: pomerium-sso-1.0.0
    app.kubernetes.io/managed-by: Tiller
  annotations:
    certmanager.k8s.io/cluster-issuer: "letsencrypt-prod"
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
    nginx.ingress.kubernetes.io/proxy-read-timeout: "600"
    
spec:
  rules:
    - host: "app.xxx.com"
      http:
        paths:
          - backend:
              serviceName: pomerium-proxy
              servicePort: https
  tls:
    - secretName: tls-prod-app
      hosts:
        - "app.xxx.com"

---
# Source: pomerium-sso/templates/sealed-secrets.yaml
apiVersion: bitnami.com/v1alpha1
kind: SealedSecret
metadata:
  name: pomerium
  labels:    
    app.kubernetes.io/name: pomerium-sso
    app.kubernetes.io/instance: pomerium
    helm.sh/chart: pomerium-sso-1.0.0
    app.kubernetes.io/managed-by: Tiller
spec:
  encryptedData:
    cookie-secret: xxx
    shared-secret: xxx
    idp-client-id: xxx
    idp-client-secret: xxx
    idp-service-account: xxx
    authenticate-cert: xxx
    authenticate-key: xxx
    authorize-cert: xxx
    authorize-key: xxx
    ca-cert: xxx
    ca-key: xxx
    proxy-cert: xxx
    proxy-key: xxx
```

**Comment by victornoel** [comment]
Here are the logs from the various services when this happens.

There is in particular an error about the service account even though I configured it, it may be why it isn't working as expected?

proxy:
```json
{"level":"info","file":"/etc/pomerium/config.yaml","time":"2019-06-03T13:28:55Z","message":"loading config from file"}
{"level":"debug","config-checksum":"68e454a078b8bf1f","time":"2019-06-03T13:28:55Z","message":"read configuration with checksum"}
{"level":"info","src":"app.xxx.com","dst":"app.app.svc.cluster.local:80","time":"2019-06-03T13:28:55Z","message":"proxy: new route"}
{"level":"debug","time":"2019-06-03T13:28:55Z","message":"proxy/clients: using a custom certificate authority"}
{"level":"debug","cert-override-name":"","addr":"https://pomerium-authenticate.proxy.svc.cluster.local","time":"2019-06-03T13:28:55Z","message":"proxy/clients: grpc connection"}
{"level":"debug","time":"2019-06-03T13:28:55Z","message":"proxy/clients: using a custom certificate authority"}
{"level":"debug","cert-override-name":"","addr":"pomerium-authorize.proxy.svc.cluster.local:443","time":"2019-06-03T13:28:55Z","message":"proxy/clients: grpc connection"}
{"level":"debug","error":"no http redirect addr provided","time":"2019-06-03T13:28:55Z","message":"cmd/pomerium: http redirect server not started"}
...
{"level":"debug","fwd_ip":"10.240.0.66","ip":"10.240.0.61","user_agent":"Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0","req_id":"7fe255a0-ab99-2e93-144a-b21c1aeb530c","error":"http: named cookie not present","time":"2019-06-03T13:30:03Z","message":"proxy: invalid session"}
{"level":"debug","fwd_ip":"10.240.0.66","ip":"10.240.0.61","user_agent":"Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0","req_id":"7fe255a0-ab99-2e93-144a-b21c1aeb530c","SigninURL":"https://auth.aks.xxx.com/sign_in?redirect_uri=https%3A%2F%2Fapp.xxx.com%2F.pomerium%2Fcallback&response_type=code&shared_secret=ZGhZZ09KVnlrK1JKMEMrbTAwSDZ5TVBNNFRkZko5dnc%3D&sig=YAkaU0p9pGniFqQ4scqrhZ2vkYajT1RJP-ebjWeK99g%3D&state=dPRtWXvSW6qRK6OB2d-QO7jVwbbvcArxzykktgRbqoUGRX1qlylSV_SIYYYrd7VZyyPi5yRK_aBKsi7quwIpvhBadi78UygGnepr-URiBARgaa-4g-uG5LJwvcccLwdrkOAOujUnBzdl5T6Fkh0DKlgnq3hxZBIMaG-0X8Gy57MgvoZ5-buKFn9xqbpcdXNfJFvZTDY&ts=1559568603","time":"2019-06-03T13:30:03Z","message":"proxy: oauth start"}
{"level":"debug","fwd_ip":"10.240.0.66","ip":"10.240.0.61","user_agent":"Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0","req_id":"7fe255a0-ab99-2e93-144a-b21c1aeb530c","duration":1.009806,"size":528,"status":302,"email":"","group":"","method":"GET","service":"proxy","url":"/","time":"2019-06-03T13:30:03Z","message":"http-request"}
```

authenticate:
```json
{"level":"info","file":"/etc/pomerium/config.yaml","time":"2019-06-03T13:28:53Z","message":"loading config from file"}
{"level":"debug","config-checksum":"33d91be925642cc4","time":"2019-06-03T13:28:53Z","message":"read configuration with checksum"}
{"level":"warn","time":"2019-06-03T13:28:53Z","message":"identity/google: no service account, cannot retrieve groups"}
{"level":"debug","error":"no http redirect addr provided","time":"2019-06-03T13:28:53Z","message":"cmd/pomerium: http redirect server not started"}
...
{"level":"debug","fwd_ip":"10.240.0.35","ip":"10.240.0.61","user_agent":"Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0","req_id":"3c2e01db-7a67-43a5-d0f2-44925f602f0f","duration":0.472802,"size":7931,"status":400,"email":"","group":"","method":"GET","service":"authenticate","url":"/sign_in?redirect_uri=https%3A%2F%2Fapp.xxx.com%2F.pomerium%2Fcallback&response_type=code&shared_secret=ZGhZZ09KVnlrK1JKMEMrbTAwSDZ5TVBNNFRkZko5dnc%3D&sig=YAkaU0p9pGniFqQ4scqrhZ2vkYajT1RJP-ebjWeK99g%3D&state=dPRtWXvSW6qRK6OB2d-QO7jVwbbvcArxzykktgRbqoUGRX1qlylSV_SIYYYrd7VZyyPi5yRK_aBKsi7quwIpvhBadi78UygGnepr-URiBARgaa-4g-uG5LJwvcccLwdrkOAOujUnBzdl5T6Fkh0DKlgnq3hxZBIMaG-0X8Gy57MgvoZ5-buKFn9xqbpcdXNfJFvZTDY&ts=1559568603","time":"2019-06-03T13:30:03Z","message":"http-request"}
```

authorize:
```json
{"level":"info","file":"/etc/pomerium/config.yaml","time":"2019-06-03T13:28:54Z","message":"loading config from file"}
{"level":"debug","config-checksum":"aec0c819fff44a1","time":"2019-06-03T13:28:54Z","message":"read configuration with checksum"}
{"level":"debug","route":"app.xxx.com","domain":"xxx.com","time":"2019-06-03T13:28:54Z","message":"add domain"}
{"level":"debug","error":"no http redirect addr provided","time":"2019-06-03T13:28:54Z","message":"cmd/pomerium: http redirect server not started"}
```

**Comment by victornoel** [comment]
Actually no, the service account error is only because the chart is not able to use a service account provided by my own secret

**Comment by desimone** [comment]
Super helpful details. I might need to chew on this a bit more but my gut is I still think this has to do with the SameSubdomain middleware validator failing. 

Are your routes (e.g)
```yml
  config.yaml: |-
    policy:
      - from: app.xxx.com
        to: http://my-app.my-app.svc.cluster.local:80
        allowed_domains:
        - xxx.com
        timeout: 10m
```
and your authenticate service /proxy service (e.g)
```
        - name: AUTHENTICATE_SERVICE_URL
          value: https://auth.aks.xxx.com
```

on the same subdomain?  For instance, the following two domains would fail to pass subdomain validation and you would see the error you are seeing

```diff
func Test_SameSubdomain(t *testing.T) {

	tests := []struct {
		name        string
		uri         string
		rootDomains string
		want        bool
	}{
		{"good url redirect", "https://example.com/redirect", "https://example.com", true},
		{"simple sub", "https://auth.example.com", "https://test.example.com", true},
		{"mismatched lengths", "https://auth.auth.example.com", "https://test.example.com", false},
+		{"does not match domain level", "https://httpbin.a.corp.example.com", "https://auth.b.corp.example.com", false},
		{"bad domain", "https://auth.example.com/redirect", "https://test.notexample.com", false},
		{"malformed url", "^example.com/redirect", "https://notexample.com", false},
		{"empty domain list", "https://example.com/redirect", ".com", false},
		{"empty domain", "https://example.com/redirect", "", false},
		{"empty url", "", "example.com", false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			u, _ := url.Parse(tt.uri)
			j, _ := url.Parse(tt.rootDomains)
			if got := SameSubdomain(u, j); got != tt.want {
				t.Errorf("SameSubdomain() = %v, want %v", got, tt.want)
			}
		})
	}
}
```

**Comment by victornoel** [comment]
@desimone they are as follows (with a, b and c different):
- app.a.xxx.com
- auth.b.c.aks.xxx.com

**Comment by victornoel** [comment]
@desimone I could try to use `auth.xxx.com` but I would prefer to keep auth to be to a lower subdomain though.

**Comment by desimone** [comment]
Yeah, I'm almost positive that's what is breaking. I believe moving authenticate's hostname to the same subdomain level would solve your immediate issue...  but that does feel like an unnecessary constraint. 

Just thinking out-loud here... I think this stems from the way oauth proxy had historically done domain checking . There used to be a `root-domain` configuration value  that was used to do validation... where as now we parse authenticate service's url and compare it to the incoming redirect url and make sure they live on the same domain.

Where things get weird is checking if a url is on the same domain is a lot trickier than it sounds because some tld's contain multiple periods (`example.co.uk`) I don't think we check based on naive pattern string matching. 

The best solution I can come up with is to use something like [EffectiveTLDPlusOne](https://godoc.org/golang.org/x/net/publicsuffix#EffectiveTLDPlusOne).

**Comment by victornoel** [comment]
@desimone and can't we just check that both domain end with the same string? :)

What I don't get is why we don't simply check that the domain of the redirect uri is one of the proxied domains? Maybe I misunderstood what is this redirect uri though :)

**Comment by desimone** [comment]
@victornoel Yes. The challenge is knowing what part of a string is a domain vs a TLD; A simple example is `example.co.uk` and `notexample.co.uk` if we naively parsed those strings we'd think they originated from the same domain. 

> What I don't get is why we don't simply check that the domain of the redirect uri is one of the proxied domains? 

Nope that's a totally valid assumption but the way the services are decoupled authenticate is agnostic to routes by design. 

At a future point I'd like to take a deeper look at what the imagined attack vector is we are guarding against with this domain check (a malicious identity provider trying to redirect a user?) but for the moment I think I have a fix #158 if you want to give that a review

**Comment by victornoel** [comment]
> @victornoel Yes. The challenge is knowing what part of a string is a domain vs a TLD; A simple example is example.co.uk and notexample.co.uk if we naively parsed those strings we'd think they originated from the same domain. Maybe there is an easier way?

Yes of course, I didn't think enough, luckily I don't work in security :P

> Nope that's a totally valid assumption but the way the services are decoupled authenticate is agnostic to routes by design.

Maybe it would make sense to re-add something like `ROOT_DOMAIN` to specify to authenticate what are the valid domains?
